README.md
[![Build Status](https://travis-ci.org/KjellMorgenstern/rbclipper.svg?branch=master)](https://travis-ci.org/KjellMorgenstern/rbclipper)


Synopsis
==========
These are Ruby bindings to Clipper, Angus Johnson's Polygon clipping
library. Because Clipper is not readily packaged, and is so beautifully
self-contained, I've included the two required files in the package.

This release contains version 6.4.2 of Clipper.

* [Clipper Homepage](http://angusj.com/delphi/clipper.php)
* [rbclipper](http://github.com/mieko/rbclipper)

To install:

    git clone git@github.com:KjellMorgenstern/rbclipper.git
    cd rbclipper
    bundle exec ruby ext/extconf.rb
    make install
    bundle exec rake test


Simple Usage:
===========
This shold be enough to get you started.  Full documentation is below.

    require 'clipper'

    a = [[0, 0], [0, 100], [100, 100], [100, 0]]
    b = [[-5, 50], [200, 50], [100, 5]]

    c = Clipper::Clipper.new

    c.add_subject_polygon(a)
    c.add_clip_polygon(b)
    c.union :non_zero, :non_zero

    => [[[100.0, 0.0], [0.0, 0.0], [0.0, 47.85714326530613], [-4.999999, 50.0],
         [0.0, 50.0], [0.0, 100.0], [100.0, 100.0], [100.0, 50.0],
         [200.0, 50.0], [100.0, 5.0]]]

Documentation
================

Clipper is a two-dimensional polygon clipping library.  `rbclipper`, the Ruby
bindings can be accessed by:

    require 'clipper'


Polygons
--------
Operations that accept or return polygons are specified as an array of `[x,y]`
coordinates, for example, to specify a triangle:

    triangle = [[0,0], [0,100], [50, -100]]

Clipper supports both holes and complex polygons.  Coordinates for output
polygons are clockwise for shells, and and counter-clockwise for holes.
See force_orientation.

Note that since 2.8, Clipper defines orientation with respect to a
_downward-increasing Y axis_, similar to how many 2D GUI/drawing APIs position
coordinate (0,0) at the top-left corner.  The bindings have followed Clipper
proper in this regard.

Multiple polygons are represented as simply an array of polygons.

Fill Types
-----------
  * `:even_odd`

    A point is considered inside the polygon if the number of edge-crossings to
    get there from outside the shape is an even number.

  * `:non_zero`

    A point is considered inside the polygon if the number of edge-crossings to
    get there is greater than zero.

  * `:positive`

    See Clipper Homepage.

  * `:negative`

    See Clipper Homepage.



Clipper::Clipper Methods
-------

* `Clipper#initialize`

   Creates a new clipper object.

* `Clipper#add_subject_polygon(polygon)`

* `Clipper#add_clip_polygon(polygon)`

  Adds a subject or clip polygon to the engine.  Boolean operations are
  calculated as `SUBJECT` *operatation* `CLIP`.  Multiple polygons can Pay attention
  to the orientation of the coordinates given; counter-clockwise for shells and
  clockwise for holes.

  Multiple subject and clip polygons can be added to the engine for operations.

* `Clipper#add_subject_polygons(expolygon)`

* `Clipper#add_clip_polygons(expolygon)`

  Add an "ExPolygon" to the engine.  Which is basically a list of polygons - the first is the
  outside (counter-clock-wise) and the rest, if any, are the holes (clock-wise).
  Boolean operations consider every expolygon added in this manner to be the same object.

* `Clipper#add_subject_polyline(polyline)`

  Add an open polygon to the engine.


* `Clipper#multiplier`

* `Clipper#multiplier=`

  Defaults to 2^10 = 1048576. Clipper since version 4.0 uses integer math instead of floating point.
  To simplify using floating point coordinates, this multiplier is multiplied to each coordinate value
  before beeing sent to Clipper, and each result coordinate is divided by the multiplier. Use 1 if you
  want to use integer coordinates.

* `Clipper#intersection(subject_fill=:even_odd, clip_fill=:even_odd, result_type=:polygons)`

* `Clipper#union(subject_fill=:even_odd, clip_fill=:even_odd, result_type=:polygons)`

* `Clipper#difference(subject_fill=:even_odd, clip_fill=:even_odd, result_type=:polygons)`

* `Clipper#xor(subject_fill=:even_odd, clip_fill=:even_odd, result_type=:polygons)`

   Performs a boolean operation on the polygons that have been added to the
   clipper object.  The result is a list of polygons or expolygons, depending on result_type being :polygons or :expolygons
   FIXME: :expolygons functionality is not implemented.

   If the result contains open polygons, they will also be returned. To see if a polygon is open, check if the first point is nil.
   FIXME: Instead of this, directly support Polytrees from Clipper

* `Clipper#offset_polygons(polygons, delta, join_type, miter_limit=0)`

  Expands the polygons by delta. Use negative delta to compress the polygons. join_type is any of :jtSquare, :jtButt, :jtMiter or :jtRound.
  Use miter_limit to make sharp joints not extend too long, by cutting off the edge.

* `Clipper#area(polygon)`

  Returns the area of the supplied polygon. The returned area is negative if the polygon points are oriented clockwise, positive otherwise. Obeys the multiplier and use_full_coordinate_range settings for the clipper object.

* `Clipper#Orientation(polygon)`

  Orientation returns a boolean value that is based on the polygon's orientation relative to the display's orientation (ie Y-axis positive upward vs Y-axis positive downward).

* On Y-axis positive upward displays, Orientation will return true if the polygon's orientation is counter-clockwise.
* On Y-axis positive downward displays, Orientation will return true if the polygon's orientation is clockwise.

Minkowski sum
-------------

    # Greek capital sigma (sum sign) ...
    sigma = [[300, 400], [100, 400], [200,300], [100,200], [300,200]]

    # diagonal brush pattern ...
    brush = [[4, -6], [6, -6], [-4, 6], [-6, 6]]

    solution = Clipper::Clipper.new.minkowski_sum(brush, [sigma], false)

    # ...

![Minkowski sum of sigma and brush](minkowski_sum.svg)

PointInPolygon
--------------

    # drop all points inside of polygon
    points.select! { |e| CLIPPER.point_in_polygon(*e, poly) == 0 }


Changelog
2018-04-22 Kjell Morgenstern
    * Upstream 6.4.2
    * Added unit tests and travis.ci
    * Added support for polylines
    * Added support for minkowski sum and diff
    * Breaking change: PtInPolygon was replaced with PointInPolygon. Now returns same values as Clipper.

2010-12-08 Mike Owens <http://mike.filespanker.com>
	* Gemification and Ruby 1.8.7 fixes courtesy of Tom Agnew
	* Upstream 2.9

2010-10-17 Mike Owens <http://mike.filespanker.com>
	* Pull upstream 2.522


test/test_helper.rb
# frozen_string_literal: true
require 'clipper'
require 'minitest/autorun'

# Note: this would be for circle ci, but we have travis
#require 'minitest-ci'

require 'minitest/junit'
# require 'minitest/reporters'

# Helper class to mange rectangles
class Rectangle
  def initialize(pos, size)
    @r = [[Float(pos[0]), Float(pos[1])], [Float(size[0]), Float(size[1])]]
  end

  def pos
    @r[0]
  end

  def pos=(val)
    @r[0] = val
  end

  def size
    @r[1]
  end

  def size=(val)
    @r[1] = val
  end

  def height
    @r[1][1]
  end

  def <=>(other)
    height * width <=> other.height * other.width
  end

  def width
    @r[1][0]
  end

  def translate_by!(delta)
    self.pos += delta
    self
  end

  def translate_by(delta)
    Rectangle.new(self.pos + delta, size)
  end

  def to_poly
    [[@r[0][0], @r[0][1]], [@r[0][0] + @r[1][0], @r[0][1]], [@r[0][0] + @r[1][0], @r[0][1] + @r[1][1]], [@r[0][0], @r[0][1] + @r[1][1]]]
  end

  def to_a
    to_poly
  end
end


test/test_clipper.rb
require 'test_helper'

class ClipperTest < Minitest::Test

  def setup
    @clip = Clipper::Clipper.new
  end

  def test_area
    r = Rectangle.new([0,0], [10, -10])
    assert_equal(@clip.area(r.to_poly), r.width * r.height)

    r = Rectangle.new([0,0], [5, 5])
    assert_equal(@clip.area(r.to_poly), r.width * r.height)
  end

  def test_orientation
    r = Rectangle.new([0,0], [10, 10])
    assert(@clip.orientation(r.to_poly))

    r = Rectangle.new([0,0], [10, -10])
    assert(not(@clip.orientation(r.to_poly)))
  end

  def test_point_in_polygon
    poly = Rectangle.new([0,0], [1.12345678, 10]).to_poly

    # point in polygon
    assert_equal(1, @clip.point_in_polygon(0.1, 0.1, poly))

    # point outside polygon
    assert_equal(0, @clip.point_in_polygon(-0.1, 0.1, poly))

    # point on polygon
    assert_equal(-1, @clip.point_in_polygon(0, 0, poly))

    # poiny on polygon 6 decimals
    assert_equal(-1, @clip.point_in_polygon(1.1234567, 5, poly))
    assert_equal(0, @clip.point_in_polygon(1.123457, 5, poly))
    assert_equal(1, @clip.point_in_polygon(1.123455, 5, poly))
  end

  def test_clip
    p1 = Rectangle.new([0,0], [2, 2]).to_poly
    p2 = Rectangle.new([0,0], [1, 1]).to_poly

    @clip.add_subject_polygon(p1)
    @clip.add_clip_polygon(p2)

    p3 = @clip.difference(:non_zero, :non_zero)

    assert_equal(1, p3.size)
    assert @clip.orientation(p3[0])
    assert_equal(3, @clip.area(p3[0]))
  end

  # FIXME: Add more tests for polylines, as this feature was disabled by default,
  # and we neglected to implement support for polytrees.
  def test_intersection_polyline
    p1 = Rectangle.new([0,0], [2, 2]).to_poly
    line = [[-1,1], [3,1]]

    @clip.add_subject_polyline(line)
    @clip.add_clip_polygon(p1)

    p3 = @clip.intersection(:non_zero, :non_zero)

    assert_equal 1, p3.size

    # Polyline is identified by nil as the first element (workaround for not handling polytrees)
    assert_nil p3[0][0]
    assert_equal [2,1], p3[0][1]
    assert_equal [0,1], p3[0][2]
  end

  def test_union
    p1 = Rectangle.new([0,0], [1, 1]).to_poly
    p2 = Rectangle.new([1,0], [1, 1]).to_poly

    @clip.add_subject_polygon(p1)
    @clip.add_subject_polygon(p2)

    p3 = @clip.union(:non_zero, :non_zero)

    assert_equal(1, p3.size)
    assert @clip.orientation(p3[0])
    assert_equal(2, @clip.area(p3[0]))
  end

  def test_minkowski_sum
    p1 = Rectangle.new([0,0], [1, 1]).to_poly
    p2 = Rectangle.new([0,0], [1, 1]).to_poly

    p3 = @clip.minkowski_sum(p1, [p2], true)

    assert_equal(1, p3.size)
    assert @clip.orientation(p3[0])
    assert_equal(4, @clip.area(p3[0]))
  end

  #FIXME: Tests go not as deep into the edgecases as they should, especially for non convex polygons
  def test_minkowski_diff
    p1 = Rectangle.new([0,0], [1, 1]).to_poly
    p2 = Rectangle.new([1,1], [0.5, 0.5]).to_poly

    p3 = @clip.minkowski_diff(p1, p2)

    assert_equal(2, p3.size)
    assert @clip.orientation(p3[0])
    assert not(@clip.orientation(p3[1]))
    assert_equal(2.25, @clip.area(p3[0]))
    assert_equal(-0.25, @clip.area(p3[1]))
  end
end


ext/clipper/rbclipper.cpp
/*
 * Clipper Ruby Bindings
 * Copyright 2010 Mike Owens <http://mike.filespanker.com/>
 * Changed by Dag Rende for Clipper after 2.9
 *
 * Released under the same terms as Clipper.
 *
 */

#include "clipper.hpp"
#include <ruby.h>

#ifndef DBL2NUM
# define DBL2NUM rb_float_new
#endif

using namespace std;
using namespace ClipperLib;

static ID id_even_odd;
static ID id_non_zero;
static ID id_positive;
static ID id_negative;
static ID id_polygons;
static ID id_ex_polygons;
static ID id_jtSquare;
static ID id_jtMiter;
static ID id_jtRound;
static ID id_etClosedPolygon;
static ID id_etClosedLine;
static ID id_etOpenButt;
static ID id_etOpenSquare;
static ID id_etOpenRound;

static inline Clipper*
XCLIPPER(VALUE x)
{
  Clipper* clipper;
  Data_Get_Struct(x, Clipper, clipper);
  return clipper;
}

static inline ClipperOffset*
XCLIPPEROFFSET(VALUE x)
{
  ClipperOffset* clipperOffset;
  Data_Get_Struct(x, ClipperOffset, clipperOffset);
  return clipperOffset;
}

static inline PolyFillType
sym_to_filltype(VALUE sym)
{
  ID inp = rb_to_id(sym);

  if (inp == id_even_odd) {
    return pftEvenOdd;
  } else if (inp == id_non_zero) {
    return pftNonZero;
  } else if (inp == id_positive) {
    return pftPositive;
  } else if (inp == id_negative) {
    return pftNegative;
  }

  rb_raise(rb_eArgError, "%s", "Expected :even_odd, :non_zero, :positive or :negative");
}

static inline EndType
sym_to_endtype(VALUE sym)
{
  ID inp = rb_to_id(sym);

  if (inp == id_etClosedPolygon) {
    return etClosedPolygon;
  } else if (inp == id_etClosedLine) {
    return etClosedLine;
  } else if (inp == id_etOpenButt) {
    return etOpenButt;
  } else if (inp == id_etOpenSquare) {
    return etOpenSquare;
  } else if (inp == id_etOpenRound) {
    return etOpenRound;
  }

  rb_raise(rb_eArgError, "%s", "Expected :etClosedPolygon, :etClosedLine or :etOpenButt or :etOpenSquare or :etOpenRound");
}

static inline JoinType
sym_to_jointype(VALUE sym)
{
  ID inp = rb_to_id(sym);

  if (inp == id_jtSquare) {
    return jtSquare;
  } else if (inp == id_jtMiter) {
    return jtMiter;
  } else if (inp == id_jtRound) {
    return jtRound;
  }

  rb_raise(rb_eArgError, "%s", "Expected :jtSquare, :jtMiter or :jtRound");
}

extern "C" {

static inline IntPoint
xy_to_intpoint(VALUE px, VALUE py, double multiplier) {
  return IntPoint((long64)(NUM2DBL(px) * multiplier), (long64)(NUM2DBL(py) * multiplier));
}

static void
ary_to_polygon(VALUE ary, ClipperLib::Path* poly, double multiplier)
{
  const char* earg =
    "Paths have format: [[p0_x, p0_y], [p1_x, p1_y], ...]";

  Check_Type(ary, T_ARRAY);

  for(long i = 0; i != RARRAY_LEN(ary); i++) {
    VALUE sub = rb_ary_entry(ary, i);
    Check_Type(sub, T_ARRAY);

    if(RARRAY_LEN(sub) != 2) {
      rb_raise(rb_eArgError, "%s", earg);
    }

    VALUE px = rb_ary_entry(sub, 0);
    VALUE py = rb_ary_entry(sub, 1);
    poly->push_back(IntPoint((long64)(NUM2DBL(px) * multiplier), (long64)(NUM2DBL(py) * multiplier)));
  }
}

static void
arys_to_polygons(VALUE polygonsValue, ClipperLib::Paths* polygons, const double multiplier)
{
  for(long i = 0; i != RARRAY_LEN(polygonsValue); i++) {
    VALUE sub = rb_ary_entry(polygonsValue, i);
    Check_Type(sub, T_ARRAY);

    ClipperLib::Path tmp;
    ary_to_polygon(sub, &tmp, multiplier);
    polygons->push_back(tmp);
  }
}

static void
rbclipper_free(void* ptr)
{
  delete (Clipper*) ptr;
}

static void
rbclipperoffset_free(void* ptr)
{
  delete (ClipperOffset*) ptr;
}

static VALUE
rbclipper_new(VALUE klass)
{
  Clipper* ptr = new Clipper;
  VALUE r = Data_Wrap_Struct(klass, 0, rbclipper_free, ptr);
  rb_obj_call_init(r, 0, 0);
  rb_iv_set(r, "@multiplier", INT2NUM(1048576));
  return r;
}

static VALUE
rbclipperoffset_new(VALUE klass)
{
  ClipperOffset* ptr = new ClipperOffset;
  VALUE r = Data_Wrap_Struct(klass, 0, rbclipperoffset_free, ptr);
  rb_obj_call_init(r, 0, 0);
  rb_iv_set(r, "@multiplier", INT2NUM(1048576));
  return r;
}

static VALUE
rbclipper_add_polygon_internal(VALUE self, VALUE polygon,
                               PolyType polytype, bool closed = true)
{
  ClipperLib::Path tmp;
  double multiplier = NUM2DBL(rb_iv_get(self, "@multiplier"));
  ary_to_polygon(polygon, &tmp, multiplier);
  XCLIPPER(self)->AddPath(tmp, polytype, closed);
  return Qnil;
}

static VALUE
rbclipper_add_polygons_internal(VALUE self, VALUE polygonsValue, PolyType polytype, bool closed = true) {
  Paths tmp;
  const double multiplier = NUM2DBL(rb_iv_get(self, "@multiplier"));
  arys_to_polygons(polygonsValue, &tmp, multiplier);
  XCLIPPER(self)->AddPaths(tmp, polytype, closed);
  return Qnil;
}

static VALUE
rbclipper_minkowski_sum(VALUE self, VALUE patternValue, VALUE pathsValue, bool isClosed = true) {
  Paths paths;
  const double multiplier = NUM2DBL(rb_iv_get(self, "@multiplier"));
  arys_to_polygons(pathsValue, &paths, multiplier);

  Path pattern;
  ary_to_polygon(patternValue, &pattern, multiplier);

  Paths solution;
  ClipperLib::MinkowskiSum(pattern, paths, solution, isClosed);

  const double inv_multiplier = 1.0 / multiplier;

  VALUE r = rb_ary_new();
  for(Paths::iterator i = solution.begin(); i != solution.end(); ++i) {
    VALUE sub = rb_ary_new();
    for(Path::iterator p = i->begin(); p != i->end(); ++p) {
        rb_ary_push(sub, rb_ary_new3(2, DBL2NUM(p->X * inv_multiplier), DBL2NUM(p->Y * inv_multiplier)));
    }
    rb_ary_push(r, sub);
  }
  return r;
}

static VALUE
rbclipper_minkowski_diff(VALUE self, VALUE poly1v, VALUE poly2v) {
  const double multiplier = NUM2DBL(rb_iv_get(self, "@multiplier"));
  Path poly1;
  ary_to_polygon(poly1v, &poly1, multiplier);
  Path poly2;
  ary_to_polygon(poly2v, &poly2, multiplier);

  Paths solution;
  ClipperLib::MinkowskiDiff(poly1, poly2, solution);

  const double inv_multiplier = 1.0 / multiplier;

  VALUE r = rb_ary_new();
  for(Paths::iterator i = solution.begin(); i != solution.end(); ++i) {
    VALUE sub = rb_ary_new();
    for(Path::iterator p = i->begin(); p != i->end(); ++p) {
        rb_ary_push(sub, rb_ary_new3(2, DBL2NUM(p->X * inv_multiplier), DBL2NUM(p->Y * inv_multiplier)));
    }
    rb_ary_push(r, sub);
  }
  return r;
}

static VALUE
rbclipper_add_subject_polyline(VALUE self, VALUE polyline)
{
  return rbclipper_add_polygon_internal(self, polyline, ptSubject, false);
}

static VALUE
rbclipper_add_subject_polylines(VALUE self, VALUE polylines)
{
  return rbclipper_add_polygons_internal(self, polylines, ptSubject, false);
}

static VALUE
rbclipper_add_subject_polygon(VALUE self, VALUE polygon)
{
  return rbclipper_add_polygon_internal(self, polygon, ptSubject);
}

static VALUE
rbclipper_add_clip_polygon(VALUE self, VALUE polygon)
{
  return rbclipper_add_polygon_internal(self, polygon, ptClip);
}

static VALUE
rbclipper_add_subject_polygons(VALUE self, VALUE polygons)
{
  return rbclipper_add_polygons_internal(self, polygons, ptSubject);
}

static VALUE
rbclipper_add_clip_polygons(VALUE self, VALUE polygons)
{
  return rbclipper_add_polygons_internal(self, polygons, ptClip);
}

static VALUE
rbclipper_clear(VALUE self)
{
  XCLIPPER(self)->Clear();
  return Qnil;
}

static VALUE
rbclipperoffset_clear(VALUE self)
{
  XCLIPPEROFFSET(self)->Clear();
  return Qnil;
}

static VALUE
rbclipper_multiplier(VALUE self)
{
  return rb_iv_get(self, "@multiplier");
}

static VALUE
rbclipper_multiplier_eq(VALUE self, VALUE multiplier)
{
  rb_iv_set(self, "@multiplier", multiplier);
  return multiplier;
}

static VALUE
rbclipper_orientation(VALUE self, VALUE polygonValue)
{
    double multiplier = NUM2DBL(rb_iv_get(self, "@multiplier"));
    ClipperLib::Path polygon;
    ary_to_polygon(polygonValue, &polygon, multiplier);

    return ClipperLib::Orientation(polygon) ? Qtrue : Qfalse;
}

static VALUE
rbclipper_area(VALUE self, VALUE polygonValue)
{
    double multiplier = NUM2DBL(rb_iv_get(self, "@multiplier"));
    ClipperLib::Path polygon;
    ary_to_polygon(polygonValue, &polygon, multiplier);

    return DBL2NUM(ClipperLib::Area(polygon) / multiplier / multiplier);
}

static VALUE
rbclipper_point_in_polygon(VALUE self, VALUE px, VALUE py, VALUE polygonValue)
{
    double multiplier = NUM2DBL(rb_iv_get(self, "@multiplier"));
    ClipperLib::Path polygon;
    ary_to_polygon(polygonValue, &polygon, multiplier);

    return INT2NUM(ClipperLib::PointInPolygon(xy_to_intpoint(px, py, multiplier), polygon));
}

static VALUE
rbclipper_offset_polygons(int argc, VALUE* argv, VALUE self)
{
    const double multiplier = NUM2DBL(rb_iv_get(self, "@multiplier"));
    const double inv_multiplier = 1.0 / multiplier;
    VALUE polygonsValue, deltaValue, joinTypeValue, endTypeValue;

    rb_scan_args(argc, argv, "31", &polygonsValue, &deltaValue, &joinTypeValue, &endTypeValue);

    Paths polygons;
    arys_to_polygons(polygonsValue, &polygons, multiplier);

    Paths resultPaths;
    XCLIPPEROFFSET(self)->AddPaths(polygons, sym_to_jointype(joinTypeValue), sym_to_endtype(endTypeValue));
    XCLIPPEROFFSET(self)->Execute(resultPaths, NUM2DBL(deltaValue) * multiplier);

    VALUE r = rb_ary_new();
    for(Paths::iterator i = resultPaths.begin(); i != resultPaths.end(); ++i) {
        VALUE sub = rb_ary_new();
        for(Path::iterator p = i->begin(); p != i->end(); ++p) {
            rb_ary_push(sub, rb_ary_new3(2, DBL2NUM(p->X * inv_multiplier), DBL2NUM(p->Y * inv_multiplier)));
        }
        rb_ary_push(r, sub);
    }
    return r;
}

static VALUE
rbclipper_execute_internal(VALUE self, ClipType cliptype,
                           VALUE subjfill, VALUE clipfill, VALUE resulttype)
{
  if (NIL_P(subjfill)) {
    subjfill = ID2SYM(id_even_odd);
  }

  if (NIL_P(clipfill)) {
      clipfill = ID2SYM(id_even_odd);
  }

  if (NIL_P(resulttype)) {
    resulttype = ID2SYM(id_polygons);
  }

  double inv_multiplier = 1.0 / NUM2LONG(rb_iv_get(self, "@multiplier"));

  VALUE r = rb_ary_new();
  if (resulttype == ID2SYM(id_polygons)) {
      PolyTree polytree;
      Paths solution;
      XCLIPPER(self)->Execute((ClipType) cliptype,
                              polytree,
                              sym_to_filltype(subjfill),
                              sym_to_filltype(clipfill));
      ClosedPathsFromPolyTree(polytree, solution);
      for(Paths::iterator i = solution.begin(); i != solution.end(); ++i) {
        VALUE sub = rb_ary_new();
        for(Path::iterator p = i->begin(); p != i->end(); ++p) {
          rb_ary_push(sub, rb_ary_new3(2, DBL2NUM(p->X * inv_multiplier), DBL2NUM(p->Y * inv_multiplier)));
        }
        rb_ary_push(r, sub);
      }
#ifdef use_lines
      OpenPathsFromPolyTree(polytree, solution);
      for(Paths::iterator i = solution.begin(); i != solution.end(); ++i) {
        VALUE sub = rb_ary_new();
        rb_ary_push(sub, Qnil);
        for(Path::iterator p = i->begin(); p != i->end(); ++p) {
          rb_ary_push(sub, rb_ary_new3(2, DBL2NUM(p->X * inv_multiplier), DBL2NUM(p->Y * inv_multiplier)));
        }
        rb_ary_push(r, sub);
      }
#endif
  }
  return r;
}

static VALUE
rbclipper_intersection(int argc, VALUE* argv, VALUE self)
{
  VALUE subjfill, clipfill, resulttype;
  rb_scan_args(argc, argv, "03", &subjfill, &clipfill, &resulttype);
  return rbclipper_execute_internal(self, ctIntersection, subjfill, clipfill, resulttype);
}

static VALUE
rbclipper_union(int argc, VALUE* argv, VALUE self)
{
  VALUE subjfill, clipfill, resulttype;
  rb_scan_args(argc, argv, "03", &subjfill, &clipfill, &resulttype);
  return rbclipper_execute_internal(self, ctUnion, subjfill, clipfill, resulttype);
}


static VALUE
rbclipper_difference(int argc, VALUE* argv, VALUE self)
{
  VALUE subjfill, clipfill, resulttype;
  rb_scan_args(argc, argv, "03", &subjfill, &clipfill, &resulttype);
  return rbclipper_execute_internal(self, ctDifference, subjfill, clipfill, resulttype);
}


static VALUE
rbclipper_xor(int argc, VALUE* argv, VALUE self)
{
  VALUE subjfill, clipfill, resulttype;
  rb_scan_args(argc, argv, "03", &subjfill, &clipfill, &resulttype);
  return rbclipper_execute_internal(self, ctXor, subjfill, clipfill, resulttype);
}

typedef VALUE (*ruby_method)(...);

void Init_clipper() {

    id_even_odd = rb_intern("even_odd");
    id_non_zero = rb_intern("non_zero");
    id_positive = rb_intern("positive");
    id_negative = rb_intern("negative");
    id_polygons = rb_intern("polygons");
    id_ex_polygons = rb_intern("expolygons");
    id_jtSquare = rb_intern("jtSquare");
    id_jtMiter = rb_intern("jtMiter");
    id_jtRound = rb_intern("jtRound");
    id_etClosedPolygon = rb_intern("etClosedPolygon");
    id_etClosedLine = rb_intern("etClosedLine");
    id_etOpenButt = rb_intern("etOpenButt");
    id_etOpenSquare = rb_intern("etOpenSquare");
    id_etOpenRound = rb_intern("etOpenRound");

    VALUE mod   = rb_define_module("Clipper");

    VALUE k = rb_define_class_under(mod, "Clipper", rb_cObject);
    rb_define_singleton_method(k, "new",
                             (ruby_method) rbclipper_new, 0);

    rb_define_method(k, "orientation",
                   (ruby_method) rbclipper_orientation, 1);
    rb_define_method(k, "point_in_polygon",
                   (ruby_method) rbclipper_point_in_polygon, 3);
    rb_define_method(k, "area",
                   (ruby_method) rbclipper_area, 1);
    rb_define_method(k, "add_subject_polygon",
                   (ruby_method) rbclipper_add_subject_polygon, 1);
    rb_define_method(k, "add_clip_polygon",
                   (ruby_method) rbclipper_add_clip_polygon, 1);
    rb_define_method(k, "add_subject_polyline",
                   (ruby_method) rbclipper_add_subject_polyline, 1);
    rb_define_method(k, "add_subject_polygons",
                   (ruby_method) rbclipper_add_subject_polygons, 1);
    rb_define_method(k, "add_clip_polygons",
                   (ruby_method) rbclipper_add_clip_polygons, 1);
    rb_define_method(k, "add_subject_polylines",
                   (ruby_method) rbclipper_add_subject_polylines, 1);
    rb_define_method(k, "clear!",
                   (ruby_method) rbclipper_clear, 0);
    rb_define_method(k, "intersection",
                   (ruby_method) rbclipper_intersection, -1);
    rb_define_method(k, "union",
                   (ruby_method) rbclipper_union, -1);
    rb_define_method(k, "difference",
                   (ruby_method) rbclipper_difference, -1);
    rb_define_method(k, "xor",
                   (ruby_method) rbclipper_xor, -1);

    rb_define_method(k, "multiplier",
                   (ruby_method) rbclipper_multiplier, 0);
    rb_define_method(k, "multiplier=",
                   (ruby_method) rbclipper_multiplier_eq, 1);

    rb_define_method(k, "minkowski_sum",
                   (ruby_method) rbclipper_minkowski_sum, 3);

    rb_define_method(k, "minkowski_diff",
                   (ruby_method) rbclipper_minkowski_diff, 2);

    VALUE ko = rb_define_class_under(mod, "ClipperOffset", rb_cObject);
    rb_define_singleton_method(ko, "new",
                             (ruby_method) rbclipperoffset_new, 0);
    rb_define_method(ko, "offset_polygons",
                  (ruby_method) rbclipper_offset_polygons, -1);
    rb_define_method(ko, "clear!",
                   (ruby_method) rbclipperoffset_clear, 0);

    rb_define_method(ko, "multiplier",
                  (ruby_method) rbclipper_multiplier, 0);
    rb_define_method(ko, "multiplier=",
                  (ruby_method) rbclipper_multiplier_eq, 1);
  }

} /* extern "C" */
